File: backend/src/models/vehicleCategory.model.js
const mongoose = require('mongoose');

// Supersedes pricingConfig.model.js's rigid 5-value vehicleType enum with an
// admin-manageable catalog: body type (top-level filter, e.g. 'open',
// 'container', 'trailer') -> sub type (the specific truck build, e.g.
// 'flat_bed', 'low_bed') -> a concrete length/tonnage/price combination.
// `vehicleType` is still the field name every other collection (Driver,
// Order) stores and matches on - it's just a stable slug now (e.g.
// 'flat_bed_20ft') instead of one of 5 hardcoded literals, so
// driver.controller.js's exact-match logic (availableOrders/acceptOrder)
// needed zero changes: a driver's registered category IS the specific
// truck spec a customer picked, same as before.
const vehicleCategorySchema = new mongoose.Schema(
  {
    vehicleType: { type: String, required: true, unique: true, trim: true },
    // Top-level filter chip in both apps' category picker.
    bodyType: { type: String, required: true, trim: true },
    // The specific build within that body type - drives which bundled
    // illustration (imageKey) and behavior a category gets.
    subType: { type: String, required: true, trim: true },
    name: { type: String, required: true, trim: true },
    lengthFt: { type: Number },
    maxWeightKg: { type: Number, required: true },
    // Maps to one of the fixed bundled illustrations in both apps'
    // assets/images/vehicles/ - not a free-text/upload field, so every
    // category stays visually consistent. See ../constants/vehicleImageKeys.js
    // for the closed set of valid values.
    imageKey: { type: String, required: true },
    baseFare: { type: Number, required: true, default: 0 },
    perKmRate: { type: Number, required: true, default: 0 },
    perKgRate: { type: Number, default: 0 },
    surgeMultiplier: { type: Number, default: 1.0 },
    isSurgeActive: { type: Boolean, default: false },
    // Whether this category carries an upfront advance at booking (both cod
    // and online) - see backend/src/utils/pricingRules.js. advanceValue
    // means % of order price (0-100) when advanceMode is 'percentage', or a
    // flat ₹ amount when 'fixed'.
    advanceRequired: { type: Boolean, default: false },
    advanceMode: { type: String, enum: ['percentage', 'fixed'], default: 'percentage' },
    advanceValue: { type: Number, default: 30 },
    // Inactive categories stay out of the customer-facing catalog and the
    // driver registration picker, but existing Drivers/Orders already using
    // that slug keep working - deactivating isn't the same as deleting.
    isActive: { type: Boolean, default: true },
    // Display order within a bodyType group in both apps' picker.
    sortOrder: { type: Number, default: 0 },
    updatedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  },
  { timestamps: true }
);

module.exports = mongoose.model('VehicleCategory', vehicleCategorySchema);

File: mobile-driver-app/lib/models/vehicle_category_model.dart
// Mirrors backend/src/models/vehicleCategory.model.js - the admin-managed
// catalog that replaced the old hardcoded 5-value vehicleType enum. Fetched
// live via GET /booking/vehicle-categories (see vehicle_config_provider.dart)
// instead of the static kVehicleTypes list this app used to ship with, so a
// new category the admin adds shows up without an app update. Same shape as
// mobile-customer-app's copy of this file - this monorepo doesn't share
// Dart code between the two apps.
class VehicleCategoryModel {
  final String vehicleType; // stable slug, e.g. 'flat_bed_20ft' - what's actually sent to the driver profile API
  final String bodyType; // top-level filter chip, e.g. 'open', 'container', 'trailer'
  final String subType; // e.g. 'flat_bed', 'low_bed' - drives which illustration to show
  final String name; // display name, e.g. "Flat Bed"
  final double? lengthFt;
  final int maxWeightKg;
  final String imageKey;

  VehicleCategoryModel({
    required this.vehicleType,
    required this.bodyType,
    required this.subType,
    required this.name,
    this.lengthFt,
    required this.maxWeightKg,
    required this.imageKey,
  });

  factory VehicleCategoryModel.fromJson(Map<String, dynamic> json) {
    return VehicleCategoryModel(
      vehicleType: json['vehicleType'] as String,
      bodyType: json['bodyType'] as String,
      subType: json['subType'] as String,
      name: json['name'] as String,
      lengthFt: (json['lengthFt'] as num?)?.toDouble(),
      maxWeightKg: (json['maxWeightKg'] as num).toInt(),
      imageKey: json['imageKey'] as String,
    );
  }

  // e.g. "Flat Bed • 20ft" or just "Bike" when there's no length.
  String get displayTitle => lengthFt != null ? '$name • ${lengthFt!.toStringAsFixed(lengthFt! % 1 == 0 ? 0 : 1)}ft' : name;

  String get weightLabel {
    if (maxWeightKg >= 1000) {
      final tons = maxWeightKg / 1000;
      return '${tons.toStringAsFixed(tons % 1 == 0 ? 0 : 1)} ton';
    }
    return 'Up to $maxWeightKg kg';
  }
}

// The closed set of bundled illustrations in assets/images/vehicles/ - must
// match backend/src/constants/vehicleImageKeys.js. 'open' is the fallback
// for any imageKey the app doesn't recognize yet (e.g. the backend catalog
// was updated with a new key before this app build shipped the matching
// asset), so a picker never renders a broken image.
const _kKnownImageKeys = {
  'bike', 'auto', 'tata_ace', 'open', 'multi_axle_open', 'container', 'flat_bed', 'low_bed', 'semi_bed',
};

String vehicleImageAsset(String imageKey) {
  final key = _kKnownImageKeys.contains(imageKey) ? imageKey : 'open';
  return 'assets/images/vehicles/$key.png';
}
